op-challenger/game/fault/types/position.go

```diff
@@ -1,56 +1,89 @@
 package types
 
-import "fmt"
+import (
+	"fmt"
+	"math/big"
+)
 
 // Position is a golang wrapper around the dispute game Position type.
 type Position struct {
 	depth        int
-	indexAtDepth int
+	indexAtDepth *big.Int
 }
 
-func NewPosition(depth, indexAtDepth int) Position {
-	return Position{depth, indexAtDepth}
+func NewPosition(depth int, indexAtDepth big.Int) Position {
+	return Position{
+		depth:        depth,
+		indexAtDepth: &indexAtDepth,
+	}
+}
+
+func NewLargePositionFromGIndex(x *big.Int) Position {
+	depth := BigMSB(x)
+	indexAtDepth := big.NewInt(0).Sub(x, big.NewInt(0).Lsh(big.NewInt(1), uint(depth)))
+	return NewPosition(depth, *indexAtDepth)
 }
 
+// todo(refcell): remove this to use the NewLargePositionFromGIndex.
 func NewPositionFromGIndex(x uint64) Position {
 	depth := MSBIndex(x)
 	indexAtDepth := ^(1 << depth) & x
-	return NewPosition(depth, int(indexAtDepth))
+	return NewPosition(depth, *big.NewInt(int64(indexAtDepth)))
+}
+
+func (p Position) Equals(other Position) bool {
+	return p.depth == other.depth && p.indexAtDepth.Cmp(other.indexAtDepth) == 0
 }
 
 func (p Position) MoveRight() Position {
 	return Position{
 		depth:        p.depth,
-		indexAtDepth: int(p.indexAtDepth + 1),
+		indexAtDepth: big.NewInt(0).Add(p.indexAtDepth, big.NewInt(1)),
 	}
 }
 
 func (p Position) Depth() int {
 	return p.depth
 }
 
-func (p Position) IndexAtDepth() int {
+func (p Position) IndexAtDepth() *big.Int {
+	if p.indexAtDepth == nil {
+		return big.NewInt(0)
+	}
 	return p.indexAtDepth
 }
 
 func (p Position) IsRootPosition() bool {
-	return p.depth == 0 && p.indexAtDepth == 0
+	return p.depth == 0 && big.NewInt(0).Cmp(p.indexAtDepth) == 0
+}
+
+func (p Position) lshIndex(amount int) *big.Int {
+	return new(big.Int).Lsh(p.IndexAtDepth(), uint(amount))
 }
 
 // TraceIndex calculates the what the index of the claim value would be inside the trace.
 // It is equivalent to going right until the final depth has been reached.
-func (p Position) TraceIndex(maxDepth int) uint64 {
+func (p Position) TraceIndex(maxDepth int) *big.Int {
 	// When we go right, we do a shift left and set the bottom bit to be 1.
 	// To do this in a single step, do all the shifts at once & or in all 1s for the bottom bits.
 	rd := maxDepth - p.depth
-	return uint64(p.indexAtDepth<<rd | ((1 << rd) - 1))
+	rhs := ((1 << rd) - 1)
```

Comment:
This still overflows for large depths:

```
func TestTraceIndexOfRootWithLargeDepth(t *testing.T) {
	traceIdx := new(big.Int).Lsh(big.NewInt(1), 100)
	pos := NewLargePositionFromGIndex(big.NewInt(1))
	actual := pos.TraceIndex(100)
	require.Equal(t, traceIdx, actual)
}
```

Not 100% my expected trace index is right, but I know the actual answer it's getting of `-1` is wrong.

Reply:
Fixed - the `rhs := ((1 << rd) - 1)` was overflowing so changed to use big integers.